### PR TITLE
Add a CapturingMatcher, for capturing mock args...

### PR DESCRIPTION
**Pull request by mmihic**
...for deeper introspection

R: @xichen2020 @robskillington

### CODE REVIEW

File: mock/capturing_matcher.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package xmock
+
+// A CapturingMatcher is a gomock.Matcher that "captures" the argument for
+// deeper introspection after the call chain completes
+type CapturingMatcher struct {
```
Comment (xichen2020) [inline mock/capturing_matcher.go]: Just curious, how is the `CapturingMatcher` going to be used? It doesn't look like the auto-generated mocks take in custom matchers, so I assume this is for hand-crafted gomock mocks?

Comment (mmihic) [inline mock/capturing_matcher.go]: autogenerated mocks take custom matchers.  cf https://github.com/m3db/m3cluster/blob/mmihic-changelists/changeset/manager_test.go#L49

Comment (xichen2020) [inline mock/capturing_matcher.go]: I see so it's passed in to methods in place of the actual arg and the captured value is used for assertions later, yeah useful for mocking out methods that take pointers etc. Sounds good.

**Comment by xichen2020** [comment]
LGTM

**Comment by robskillington** [comment]
You can also use the `.Do(func(arg1 arg1type, arg2 arg2type) { capturedArg1 = arg1 })` to capture the argument instead of a matcher btw.  In fact that might be cleaner than creating a matcher that just captures the argument as a matcher is orthogonal to capturing data from the call.

**Comment by kobolog** [comment]
This PR is pretty much stale, are we going to merge it or should we just abandon it? Let's not let PRs to hang around w/o any activity for more than a few weeks.

**Pull request closed without merging**